### PR TITLE
Performance improvements for Proxy::invokeDefaultMethod

### PR DESCRIPTION
**Pull request by plevart**
This patch makes the Proxy::invokeDefaultMethod spec about thrown exception types consistent with Method::invoke spec

### CODE REVIEW

File: src/java.base/share/classes/java/lang/reflect/Proxy.java [hunk 1]
```diff
@@ -1270,61 +1283,63 @@ public static Object invokeDefaultMethod(Object proxy, Method method, Object...
...
             if (cached != null) {
                 superMH = cached;
-            }
-        }
-
-        // validate the arguments if they match the method type
-        int numArgs = args != null ? args.length : 0;
-        if (numArgs != type.parameterCount()) {
-            throw new IllegalArgumentException("args not matching the formal parameter types: " + type);
-        }
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/reflect/Proxy.java:1302]: We can retain this parameter count check as a quick validation.

File: src/java.base/share/classes/java/lang/reflect/Proxy.java [hunk 2]
```diff
@@ -1241,21 +1245,30 @@ public static InvocationHandler getInvocationHandler(Object proxy)
...
      * @jvms 5.4.3. Method Resolution
      */
     public static Object invokeDefaultMethod(Object proxy, Method method, Object... args)
             throws InvocationTargetException {
         Objects.requireNonNull(proxy);
         Objects.requireNonNull(method);
+        if (args == null) {
+            // consistency with Method::invoke: null args array is equivalent to empty array
+            args = EMPTY_ARGS;
+        }
```
Comment (mlchung) [inline src/java.base/share/classes/java/lang/reflect/Proxy.java:1271]: alternatively, move to this line 1331: `superMH.invokeExact(proxy, (args != null ? args : EMPTY_ARGS));

File: test/jdk/java/lang/reflect/Proxy/DefaultMethods.java [hunk 3]
```diff
@@ -322,33 +322,31 @@ public void testIllegalArgument(Class<?>[] interfaces, Class<?> defc, String nam
...
 
-    @Test(expectedExceptions = {IllegalArgumentException.class})
-    public void testNullArguments() throws Exception {
-        ClassLoader loader = DefaultMethods.class.getClassLoader();
-        I4 proxy = (I4)Proxy.newProxyInstance(loader, new Class<?>[]{I4.class}, HANDLER);
-        Method m = I4.class.getMethod("mix", int.class, String.class);
-        assertTrue(m.isDefault());
-        Proxy.invokeDefaultMethod(proxy, m, (Object[])null);
-    }
-
```
Comment (mlchung) [inline test/jdk/java/lang/reflect/Proxy/DefaultMethods.java:351]: Do you mind adding a new test case to check if NPE thrown by the method body resulting in `InvocationTargetException`?   I missed that case.

File: test/jdk/java/lang/reflect/Proxy/DefaultMethods.java [hunk 4]
```diff
@@ -322,33 +322,31 @@ public void testIllegalArgument(Class<?>[] interfaces, Class<?> defc, String nam
...
 
-    @DataProvider(name = "arguments")
-    private Object[][] arguments() {
+    @DataProvider(name = "illegalArguments")
+    private Object[][] illegalArguments() {
         return new Object[][] {
-                new Object[] {},
-                new Object[] { 100, "foo", 100 },
-                new Object[] { "foo", 100},
+            new Object[] {},
```
Comment (mlchung) [inline test/jdk/java/lang/reflect/Proxy/DefaultMethods.java:336]: I'm okay with using 4-space indentation here.  Would you mind changing other data providers to use the same format consistently?

**Review by mlchung — APPROVED** [review]
Thanks for doing this.  Looks good.

As for the spec change, I may tweak the wording a little once you push it to my branch.

File: src/java.base/share/classes/java/lang/reflect/Proxy.java (hunk 1 above)
Comment (plevart) [reply src/java.base/share/classes/java/lang/reflect/Proxy.java:1302]: Unless you really want to customize the message (perhaps to say more directly what is wrong - not the types but the number of parameters) then this is not really needed as the same check is performed later in MH.invokeExact and the same type of exception is thrown. The message of the IAE thrown by MH chain is: "array is not of length %d" - I think this is almost on the spot. WDYT?

File: src/java.base/share/classes/java/lang/reflect/Proxy.java (hunk 2 above)
Comment (plevart) [reply src/java.base/share/classes/java/lang/reflect/Proxy.java:1271]: I did exactly like you're proposing, but wrong code got generated. invokeExact is a @PolymorphicSignature method with varargs. javac treats it differently and when a local var (method parameter) is passed to it, javac correctly generates the invocation but when a ? : expression is passed to it, javac does it differently. Perhaps this is a bug in javac.

File: test/jdk/java/lang/reflect/Proxy/DefaultMethods.java (hunk 3 above)
Comment (plevart) [reply test/jdk/java/lang/reflect/Proxy/DefaultMethods.java:351]: Ok.

File: test/jdk/java/lang/reflect/Proxy/DefaultMethods.java (hunk 4 above)
Comment (plevart) [reply test/jdk/java/lang/reflect/Proxy/DefaultMethods.java:336]: Ok.

Comment (plevart) [reply test/jdk/java/lang/reflect/Proxy/DefaultMethods.java:336]: I added a commit for indentation and new test case in test. I can't directly push this to your repository:

> ERROR: Permission to mlchung/jdk.git denied to plevart

But you can merge this PR into your branch and the result will be the same: Your PR against jdk/master will be updated with merged changes from my branch.

**Pull request merged**